File: EJERCICIOS/6.py
"""
Un alumno desea saber cual sera su promedio general
en tres materias mas dificiles que cursa y cual sera
el promedio que obtendra en cada una de ellas,
estas materias se evaluan como se muestra a continuacion
1. Las calificaciones de programacion se obtienen 
de la siguiente manera:
    Examen: 90%, promedio de las practicas: 10% y se pidieron
    un total de 2 practicas
2. Las calificaciones de algebra lineal se obtienen 
de la siguiente manera:
    Examen: 80%, promedio de las practicas: 20% y se pidieron
    un total de 2 practicas
3. Las calificaciones de Ing de sistemas se obtienen 
de la siguiente manera:
    Examen: 85%, promedio de las practicas: 15% y se pidieron
    un total de 2 practicas
Ejemplo de diccionario = {"asignatura":"hola","examen":10,"practica1":50,"practica2":60}
"""



def ingresarDatos():
    asignaturas=("Programacion","Algebra Lineal","Ing de Sistemas")
    datos = []
    for asignatura in asignaturas:
        print(f"Ingrese la nota de examen de {asignatura}")
        examen = float(input())
        print(f"Ingrese la nota de practica 1  de {asignatura}")
        practica1 = float(input())
        print(f"Ingrese la nota de practica 2 de {asignatura}")
        practica2 = float(input())
        datos.append({"asignatura":asignatura,"examen":examen,
                      "practica1":practica1,"practica2":practica2})
    return datos

def calcular():
    datos = ingresarDatos()
    promedios = []
    for dato in datos:
        if dato.get("asignatura")=="Programacion":
            promedio = dato.get("examen")*0.9+(dato.get("practica1")+dato.get("practica2")/2)*0.1
        elif dato.get("asignatura")=="Algebra Lineal":
            promedio = dato.get("examen")*0.8+(dato.get("practica1")+dato.get("practica2")/2)*0.2
        elif dato.get("asignatura")=="Ing de Sistemas":
            promedio = dato.get("examen")*0.85+(dato.get("practica1")+dato.get("practica2")/2)*0.15
        promedios.append({"asignatura":dato.get("asignatura"),"promedio":promedio})
    return promedios
#Hacer la funcion mostrar promedios#
calcular()

File: EJERCICIOS/typescript/6.ts
import readline from 'node:readline/promises';
import { stdin as input, stdout as output } from 'node:process';

const rl = readline.createInterface({ input, output });

type NotaMateria = {
  asignatura: string;
  examen: number;
  practica1: number;
  practica2: number;
};

type PromedioMateria = {
  asignatura: string;
  promedio: number;
};

async function ingresarDatos(): Promise<NotaMateria[]> {
  const asignaturas = ['Programacion', 'Algebra Lineal', 'Ing de Sistemas'];
  const datos: NotaMateria[] = [];

  for (const asignatura of asignaturas) {
    console.log(`Ingrese la nota de examen de ${asignatura}`);
    const examen = Number(await rl.question(''));

    console.log(`Ingrese la nota de practica 1 de ${asignatura}`);
    const practica1 = Number(await rl.question(''));

    console.log(`Ingrese la nota de practica 2 de ${asignatura}`);
    const practica2 = Number(await rl.question(''));

    datos.push({ asignatura, examen, practica1, practica2 });
  }

  return datos;
}

async function calcular(): Promise<PromedioMateria[]> {
  const datos = await ingresarDatos();
  const promedios: PromedioMateria[] = [];

  for (const dato of datos) {
    const promedioPracticas = (dato.practica1 + dato.practica2) / 2;
    let promedio = 0;

    if (dato.asignatura === 'Programacion') {
      promedio = dato.examen * 0.9 + promedioPracticas * 0.1;
    } else if (dato.asignatura === 'Algebra Lineal') {
      promedio = dato.examen * 0.8 + promedioPracticas * 0.2;
    } else if (dato.asignatura === 'Ing de Sistemas') {
      promedio = dato.examen * 0.85 + promedioPracticas * 0.15;
    }

    promedios.push({ asignatura: dato.asignatura, promedio });
  }

  return promedios;
}

function mostrarPromedios(promedios: PromedioMateria[]): void {
  let promedioGeneral = 0;

  for (const dato of promedios) {
    console.log(`Promedio de ${dato.asignatura}: ${dato.promedio}`);
    promedioGeneral += dato.promedio;
  }

  console.log(`Promedio general: ${promedioGeneral / promedios.length}`);
}

const promedios = await calcular();
mostrarPromedios(promedios);
rl.close();

